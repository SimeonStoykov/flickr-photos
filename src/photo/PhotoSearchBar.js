import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import {
    fetchPhotos,
    setSearchInputValue,
    setSearchTerm,
    setCurrentPage
} from '../photo/PhotoActions';

export class PhotoSearchBar extends Component {
    constructor() {
        super();
        this.searchPhotos = this.searchPhotos.bind(this);
        this.handleSearchInputValueChange = this.handleSearchInputValueChange.bind(this);
        this.handleKeyPress = this.handleKeyPress.bind(this);
    }

    handleSearchInputValueChange(e) {
        this.props.setSearchInputValue(e.target.value);
    }

    handleKeyPress(e) {
        if (e.key === 'Enter' && this.props.searchInputValue) {
            this.searchPhotos();
        }
    }

    searchPhotos() {
        let { setCurrentPage, setSearchTerm, searchInputValue, fetchPhotos } = this.props;
        const newSearchTerm = searchInputValue.trim();
        setSearchTerm(newSearchTerm);
        setCurrentPage(1);
        fetchPhotos({ currentPage: 1, searchTerm: newSearchTerm });
    }

    render() {
        return (
            <div className="header--search-bar">
                <input type="text" value={this.props.searchInputValue} onChange={this.handleSearchInputValueChange} onKeyPress={this.handleKeyPress} placeholder="Search in title, description and tags..." />
                <button onClick={this.searchPhotos} disabled={!this.props.searchInputValue}></button>
            </div>
        );
    }
}

PhotoSearchBar.propTypes = {
    fetchPhotos: PropTypes.func,
    searchTerm: PropTypes.string,
    setSearchTerm: PropTypes.func,
    setCurrentPage: PropTypes.func,
    currentPage: PropTypes.number
};

const mapStateToProps = ({ photosData }) => {
    return {
        searchInputValue: photosData.get('searchInputValue'),
        searchTerm: photosData.get('searchTerm'),
        currentPage: photosData.get('currentPage')
    };
}

const mapDispatchToProps = dispatch => {
    return {
        fetchPhotos: dataUrl => dispatch(fetchPhotos(dataUrl)),
        setSearchInputValue: value => dispatch(setSearchInputValue(value)),
        setSearchTerm: value => dispatch(setSearchTerm(value)),
        setCurrentPage: value => dispatch(setCurrentPage(value))
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(PhotoSearchBar);
