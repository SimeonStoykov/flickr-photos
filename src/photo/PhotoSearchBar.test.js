import React from 'react';
import { shallow } from 'enzyme';
import '../setupTests';
import { PhotoSearchBar } from './PhotoSearchBar';

describe('Photo Search Bar', () => {
    let component;
    const mockSetSearchInputValueFn = jest.fn();

    beforeEach(() => {
        component = shallow(<PhotoSearchBar setSearchInputValue={mockSetSearchInputValueFn} />)
    });

    it('should render search bar correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('should call setSearchInputValue on input value change', () => {
        component.find('input[type="text"]').simulate('change', { target: { value: 'cats' } });
        expect(mockSetSearchInputValueFn).toHaveBeenCalled();
    });
});