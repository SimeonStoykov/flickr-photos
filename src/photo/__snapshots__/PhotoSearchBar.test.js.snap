// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo Search Bar should render search bar correctly 1`] = `
<div
  className="header--search-bar"
>
  <input
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Search in title, description and tags..."
    type="text"
  />
  <button
    disabled={true}
    onClick={[Function]}
  />
</div>
`;
